tests/test_lion8b.py

```diff
@@ -0,0 +1,549 @@
+        mom_new = d_new['exp_avg']
+
+        assert mom_orig.shape == mom_new.shape
+        assert mom_orig.dtype == mom_new.dtype
+        if use_errors:
+            errs_orig = d_orig['errors']
+            errs_new = d_new['errors']
+            assert errs_orig.shape == errs_new.shape
+            assert errs_orig.dtype == errs_new.dtype
+
+        if state_sharding != _FULL_STATE:
+            continue  # more detailed checks lean on FSDP impl details
+
+        # momentums may not be bit-for-bit identical because Optimizer upcasts
+        # to f32 and we convert back to bf16, possibly with different rounding
+        torch.testing.assert_close(mom_orig, mom_new)
+        # errors not bit-for-bit identical because scales get upcast too
+        if use_errors and (dtype != torch.float32):
+            torch.testing.assert_close(d_orig['errors'], d_new['errors'])
+
+
+@pytest.mark.gpu
+@pytest.mark.parametrize('N,D', [(32, 32), (256, 256), (1024, 1024),
+                                 (4096, 4096), [16384, 16384]])
+def test_fused_as_fast_as_unfused(N: int,
+                                  D: int,
+                                  min_elems_traversed: int = 1000000):
+    W = torch.randn((N, D), device='cuda', requires_grad=True)
+    W.grad = torch.randn((N, D), device='cuda', requires_grad=False)
+
+    num_iters = int(np.ceil(min_elems_traversed / W.grad.numel()))
+    num_iters = min(100, num_iters)  # don't take all day when overhead-bound
+
+    times = {}
+    kwargs = {'weight_decay': .01}
+    combos = [(True, False), (True, True), (False, False), ('NA', False)]
+    for fused, use_errors in combos:
+        if fused == 'NA':
+            opt = Lion8bit([W], quantize=False,
+                           **kwargs)  # type:ignore (reportGeneralTypeIssues)
+        else:
+            opt = Lion8bit([W],
+                           _fused=fused,
+                           error_correction=use_errors,
+                           **kwargs)  # type:ignore (reportGeneralTypeIssues)
+        for _ in range(3):
+            opt.step()  # warmup iters
+        torch.cuda.synchronize()
+        t_start = time.time()
+        for _ in range(num_iters):
+            opt.step()
+        torch.cuda.synchronize()
+        t_end = time.time()
+        dur = (t_end - t_start) / num_iters
+        if use_errors:
+            times['ecc'] = dur
+        else:
+            times[fused] = dur
+
+    atol = 20e-6  # should always be faster, but avoids rare flakiness
+    assert times[True] < times[False] + atol
+    assert times[True] < times['NA'] + atol
+    assert times['ecc'] < times['NA'] + atol
+
+    if False:  # change to True to check on thruput
+        print('')
+        print('time fused (ms):       ', times[True] * 1e3)
+        print('time fused+ecc (ms):   ', times['ecc'] * 1e3)
+        print('time unfused (ms):     ', times[False] * 1e3)
+        print('time unquantized (ms): ', times['NA'] * 1e3)
```

Comment:
iirc pytest doesn't print stuff out unless you add a  -s flag. ie just printing this probably wont hurt

Reply:
done